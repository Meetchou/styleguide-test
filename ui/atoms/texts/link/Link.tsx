import React, { MouseEvent, useMemo } from 'react';
import cls from 'classnames';

export type LinkProps = {
  children: React.ReactNode;
  onClick?: (event: MouseEvent, href: string | undefined) => void;
  href?: string;
  target?: string;
  className?: string;
  isBlock?: boolean;
};

const Link:React.ForwardRefRenderFunction<HTMLAnchorElement, LinkProps> = ({
  children,
  className,
  href,
  onClick,
  isBlock,
  ...rest
}, ref) => {
  const handleClick = useMemo(
    () => {
      if (!onClick) {
        return undefined;
      }

      return (e: MouseEvent): void => {
        const { shiftKey, altKey, ctrlKey, metaKey } = e;
        if (!(shiftKey || altKey || ctrlKey || metaKey)) {
          e.preventDefault();
          onClick(e, href);
        }
      };
    },
    [onClick, href]
  );

  return (
    <a
      {...rest}
      ref={ref}
      onClick={handleClick}
      href={href}
      className={cls(
        'text-primary-500',
        (onClick || href) && !isBlock ? 'hover:underline' : '',
        onClick || href ? 'cursor-pointer' : 'cursor-default',
        className
      )}
    >
      {children}
    </a>
  );
};

export default React.forwardRef(Link);
