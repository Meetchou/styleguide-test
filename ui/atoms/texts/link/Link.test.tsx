import React from 'react';

import render from '@ui/testing/render';

import Link from './Link';

const actionClasses = ['cursor-pointer', 'hover:underline'];

describe('atoms/texts/link/Link', () => {
  it('renders correctly', () => {
    const tree = render(
      <Link
        href='http://perdu.com'
        onClick={jest.fn()}
      >
        test
      </Link>
    );

    expect(tree.toJSON()).toMatchSnapshot();

    const { className } = tree.root.findByType('a').props;
    actionClasses.forEach(
      name => expect(className).toContain(name)
    );
  });

  it('renders correctly with isBlock', () => {
    const tree = render(
      <Link
        href='http://perdu.com'
        onClick={jest.fn()}
        isBlock
      >
        test
      </Link>
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });

  describe('onClick action', () => {
    describe('should not be fired when click is made with key:', () => {
      it.each([
        'shift',
        'ctrl',
        'meta',
        'alt'
      ])('%s', key => {
        const props = {
          onClick: jest.fn(),
          href: 'http://perdu.com'
        };
        const tree = render(<Link {...props} > test </Link>);

        const event = { [`${key}Key`]: true };
        tree.root.findByType('a').props.onClick(event);

        expect(props.onClick).not.toHaveBeenCalled();
      });
    });

    describe('should be fired when click is made', () => {
      it('with no key', () => {
        const props = {
          onClick: jest.fn(),
          href: 'http://perdu.com'
        };
        const tree = render(<Link {...props} > test </Link>);

        const event = {
          testKey: true,
          preventDefault: jest.fn()
        };
        tree.root.findByType('a').props.onClick(event);

        expect(props.onClick).toHaveBeenCalledWith(event, props.href);
        expect(event.preventDefault).toHaveBeenCalled();
      });

      it('with no special key', () => {
        const props = {
          onClick: jest.fn(),
          href: 'http://perdu.com'
        };
        const tree = render(<Link {...props} > test </Link>);

        const event = {
          testKey: true,
          preventDefault: jest.fn()
        };
        tree.root.findByType('a').props.onClick(event);

        expect(props.onClick).toHaveBeenCalledWith(event, props.href);
        expect(event.preventDefault).toHaveBeenCalled();
      });
    });
  });

  it('should not have cursor and decoration when onClick and href missing', () => {
    const tree = render(<Link>test</Link>);

    const { className } = tree.root.findByType('a').props;
    actionClasses.forEach(
      name => expect(className).not.toContain(name)
    );
  });
});
