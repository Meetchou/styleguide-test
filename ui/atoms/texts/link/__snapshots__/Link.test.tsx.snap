// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atoms/texts/link/Link renders correctly 1`] = `
<a
  className="text-primary-500 hover:underline cursor-pointer"
  href="http://perdu.com"
  onClick={[Function]}
>
  test
</a>
`;

exports[`atoms/texts/link/Link renders correctly with isBlock 1`] = `
<a
  className="text-primary-500 cursor-pointer"
  href="http://perdu.com"
  onClick={[Function]}
>
  test
</a>
`;
